core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java

```diff
@@ -117,6 +156,32 @@ public void setDisposer(@NonNull Disposer disposer) {
      * Must be safely serializable, so it receives runtime context comparable to that of the original setup.
      */
     public static abstract class Disposer implements Serializable {
+
+        /**
+         * Creates a new end-of-wrapped-block callback.
+         *
+         * @deprecated Use the constructor that takes a wrapper instance instead.
+         */
+        @Deprecated
+        protected Disposer() {
+            this.requiresWorkspace = true;
+        }
+
+        /**
+         * Creates a new end-of-wrapped-block callback.
+         *
+         * @param wrapper The wrapper for which this is a callback; used to determine whether or not this callback
+         *                requires a workspace context (working directory and launcher). That choice determines whether
+         *                {@link #tearDown(Run, FilePath, Launcher, TaskListener)} or
+         *                {@link #tearDown(Run, TaskListener)} applies.
+         * @since TODO
+         */
+        protected Disposer(@NonNull SimpleBuildWrapper wrapper) {
+            this.requiresWorkspace = wrapper.requiresWorkspace();
+        }
+
+        private final boolean requiresWorkspace;
```

Comment:
Invert the sense of the field since a deserialized `Disposer` from a Jenkins session prior to upgrade (for Pipeline builds) will have fields set to default values like `false` unless you implement `readResolve`.

Reply:
Do these actually get persisted? I thought they were only serialized for sending to the agent for execution.

But sure, inverting the field's meaning is easy.
Or, I could just make this a method like on the other objects, avoiding the serialization issue entirely.

Reply:
`Disposer`s are _not_ sent to the agent. They are serialized in a Pipeline build, so that you can write for example

```groovy
xvnc {
  sh 'make world'
}
```

and have Jenkins restart in the middle of the `make` command but still know what to clean up later.

Reply:
> a method like on the other objects

As noted in the other PR, I would prefer to have a single flag for an implementor that covers both setup and teardown.

Reply:
I know - but unfortunately neither the creation nor the execution of a `Disposer` is directly tied to a `SimpleBuildWrapper`.
Unless it would be acceptable to add a constructor to `Disposer` that takes a `SimpleBuildWrapper`; then it could set its internal flag from that.

Reply:
see https://github.com/jenkinsci/jenkins/pull/4829#discussion_r450575943

Reply:
OK, will look at how to arrange it like that.